### DISCUSSION TITLE
filter() type not properly derived from overriden __iter__()

### DISCUSSION THREAD

**Discussion by Fatsie** [microsoft/pylance-release#2709] [Enhancement]
## Environment data

- Language Server version: 2021.5.1 (pyright df8b6b3c)
- OS and version: Ubuntu 20.04
- Python version (& distribution if applicable, e.g. Anaconda): Python 3.7.9 64-bit (conda)

## Code Snippet

```python
from typing import Generic, Iterable, Iterator, Mapping, Sequence, Union

class A:
    def __init__(self, idx: str, v: float):
        self.idx = idx
        self.v = v

class SequenceMapping(Mapping[str, A], Sequence[A]):
    def __init__(self, iterable: Iterable[A]=tuple()):
        self._list = list(iterable)
        self._map = {e.idx: e for e in self._list}

    def __getitem__(self, key: Union[int, str]) -> A: # type: ignore[override]
        if isinstance(key, int):
            return self._list[key]
        elif isinstance(key, str):
            return self._map[key]
    
    def __iter__(self) -> Iterator[A]: # type: ignore[override]
        return iter(self._list)

    def __len__(self) -> int:
        return len(self._list)

l = SequenceMapping((A("zero", 0.0),))

for a1 in l:
    print(a1.v)

for a2 in filter(lambda _: True, l):
    print(a2.v)
```

## Expected behaviour

Not give a mypy/Pylance error

## Actual behaviour

First `for` loop has proper type A for a1, second for loop has wrongly derived type `str` for a2. Running the script gives the right results. The wrong type for a2 gives a mypy and Pylance error:

mypy:
```
{
	"resource": "/home/verhaegs/eda/test/test_filter.py",
	"owner": "python",
	"code": "error",
	"severity": 8,
	"message": "\"str\" has no attribute \"v\"",
	"source": "mypy",
	"startLineNumber": 31,
	"startColumn": 11,
	"endLineNumber": 31,
	"endColumn": 11
}
```

Pylance:
```
{
	"resource": "/home/verhaegs/eda/test/test_filter.py",
	"owner": "_generated_diagnostic_collection_name_#2",
	"code": {
		"value": "reportGeneralTypeIssues",
		"target": {
			"$mid": 1,
			"external": "https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules",
			"path": "/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md",
			"scheme": "https",
			"authority": "github.com",
			"fragment": "diagnostic-severity-rules"
		}
	},
	"severity": 8,
	"message": "Cannot access member \"v\" for type \"str\"\n  Member \"v\" is unknown",
	"source": "Pylance",
	"startLineNumber": 31,
	"startColumn": 14,
	"endLineNumber": 31,
	"endColumn": 15
}```

**Comment by erictraut** [comment]
The reason you're seeing this in pylance (and probably in mypy as well) is due to an important optimization in the type analyzer. When it performs type compatibility analysis for a protocol class (such as `Iterable`), it first checks to see if the class explicitly derives from that protocol class. If so, it skips the expensive member-by-member compatibility check that is otherwise required for protocol verification. In this example, `SequenceMapping` derives from two classes that both implement the `Iterable` protocol. The `Mapping` class iterates over its keys (in this case, `str`), and the `Sequence` class iterates over its values (in this case, `A`). Your class implements the `__iter__` method and iterates over `A`, so it's acting as though it's a `Sequence` rather than a `Mapping`.

Fixing this in the type analyzer would require the removal of this very important optimization, slowing down analysis in a measurable way for everyone. Since this is such an edge case (the first time anyone has reported it for pylance or mypy, to my knowledge), I'm not inclined to make that change.

There is a simple workaround that you can apply: simply swap `Mapping` and `Sequence` in the class declaration for `SequenceMapping`. By placing `Sequence` first, the MRO (member resolution ordering) will use the `Sequence` behavior of `__iter__`, which is compatible with the implementation in `SequenceMapping`. I've confirmed that this workaround works for both pylance and mypy.

**Comment by Fatsie** [comment]
What is the difference then between the first and second for loop so that he can do it right in the first one and not in the second one?

**Comment by erictraut** [comment]
The second loop uses `filter`, which is a generic class whose constructor is defined as follows in the type stub:
```python
    def __init__(self, __function: Callable[[_T], Any], __iterable: Iterable[_T]) -> None: ...
```
To determine the type argument for `filter`, the type constraint solver needs to solve for `_T` by assigning the type of argument `l` to the parameter type `Iterable[_T]`. In the process, it needs to verify that `l`'s type conforms to the `Iterable` protocol. That's the expensive part that is skipped when the class of the argument type is an explicit subclass of `Iterable`, as it is in this example.

In the first loop, the type checker directly looks up `__iter__`. There's no need to verify that it conforms to a protocol in this case.

**Comment by Fatsie** [comment]
OK, but then I think it should still be able to derive the right type of `Iterable` by looking up `__iter__` of `l` also in the second loop to see if it is overloaded. Is that so expensive ?

**Comment by erictraut** [comment]
The type checker doesn't know anything special about the symbol `Iterable`. It knows that it's a protocol class and must compare the types of all members within it. For `Iterable` this isn't very expensive because it has only one member defined, but most other protocol classes are more complex and may have dozens of members.

**Comment by Fatsie** [comment]
@savannahostrowski It's not clear to me that the 'waiting for info' is meant that I have to give feedback.
Anyway, as a user implementation details are not something that plays a role in seeing something as a bug.
As a user I do see that you can override methods with incompatible signature using '# type: ignore[override]' annotation.
I now see that the Pylance type derivation may sometimes ignore this annotation; seemingly for speed reasons. I am not convinced this is only for esoteric edge cases.

**Comment by erictraut** [comment]
Did you try the workaround that I suggested above? If so, did that address the problem for you?

The reason I suspect you are the first (and perhaps only) person to hit this issue is that it's very rare to inherit from two classes that both implement protocols that overlap in incompatible ways. Furthermore, I don't think anyone has filed a bug against mypy in its many-year history related to this behavior.

I think it's fine to keep this issue open. If others encounter it, they should feel free to upvote it.

**Comment by savannahostrowski** [comment]
@fatsie Thanks for the feedback regarding the tag. I've updated it so that it's less ambiguous.

**Comment by Fatsie** [comment]
I did already work around the issue but I prefer to not have to need to know the imperfections of a tool that I need to work around.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Reply by erictraut** [reply]
I've added a new feature in pyright to detect cases where class is defined with mutually-incompatible base classes. This should be considered an error (and flagged as such by a type checker) because it leads to ambiguous situations. For example, if we look at the code snippet that started this thread, `SequenceMapping` has two base classes: `Mapping[str, A]` and `Sequence[A]`. These are inconsistent because `Mapping[str, A]` is a subclass of `Sequence[str]` which isn't compatible with `Sequence[A]`. Pyright now flags this as an error. This will be included in the next release of pyright and the next pre-rerelease version of pylance.
